torch/csrc/autograd/python_variable.cpp

```diff
@@ -137,28 +137,49 @@ PyObject * THPVariable_Wrap(Variable var)
 static int THPVariable_traverse(THPVariable *self, visitproc visit, void *arg)
 {
   Py_VISIT(self->backward_hooks);
-  // We don't want to traverse the grad_fn, even if the Variable owns it and the
-  // shared pointer's use count is 1. This is because we would need to treat
-  // the grad_fn as part of the Python state and hold the GIL sometimes when
-  // grad_fn's shared_ptr is copied, otherwise a race condition with the Python
-  // GC could occur. Holding the GIL when the shared_ptr is copied adds
-  // undesirable complexity/overhead.
-  //
-  // When hooks, a Variable, and its grad_fn are involved in a Python reference
-  // cycle, because we're not traversing the grad_fn, the reference cycle will
-  // in fact leak.
-  //
-  // See https://gist.github.com/zou3519/7ac92b84dd7d206dcc6eae55fee8372c
-  // for more details about the race condition involving traversing the grad_fn
-  // and the python GC.
+
   const auto& tensor = THPVariable_Unpack(self);
   if (tensor.defined()) {
+
+    // WARNING: The grad_fn traversal logic is very subtle, if you change this,
+    // be very careful not to re-introduce this bug:
+    // https://gist.github.com/zou3519/7ac92b84dd7d206dcc6eae55fee8372c
+
+    // We want to be conservative here and make sure that we only traverse the
+    // grad_fn if this python object is the sole owner of the grad_fn and no
+    // other thread can create another owner without taking the GIL first.
+    // This is because we need to ensure that the "owning or not" of the grad_fn
+    // doesn't change between this traverse and the clear below.
```

Comment:
It's worth noting that this assumption is only valid if you don't share non-owning reference to tensor across thread. Which is crazy dangerous but could be valid if you did something like block on the other thread before deallocating orig tensor. I don't care but we do need to make sure people don't do crazy shit like this.